File: backend/app/services/embeddings_service.py
from __future__ import annotations

import logging
from typing import Any

import httpx
from sqlalchemy.orm import Session

from app.core.config import settings
from app.core.crypto import secret_manager
from app.models.asset_chunk import AssetChunk
from app.models.provider import Provider
from app.services.workspace_service import WorkspaceService

logger = logging.getLogger(__name__)

OPENAI_COMPATIBLE_PROVIDER_TYPES = ("openai", "openrouter", "litellm", "local")
SUPPORTED_EMBEDDING_PROVIDER_TYPES = (*OPENAI_COMPATIBLE_PROVIDER_TYPES, "gemini", "ollama")
EMBEDDING_PROVIDER_PRIORITY = {
    "openai": 0,
    "openrouter": 1,
    "gemini": 2,
    "ollama": 3,
    "litellm": 4,
}
DEFAULT_BASE_URLS = {
    "openai": "https://api.openai.com/v1",
    "openrouter": "https://openrouter.ai/api/v1",
    "litellm": "http://localhost:4000/v1",
    "local": "http://localhost:4000/v1",
    "gemini": "https://generativelanguage.googleapis.com/v1beta",
    "ollama": "http://localhost:11434",
}
DEFAULT_EMBEDDING_MODELS = {
    "openai": "text-embedding-3-small",
    "openrouter": "openai/text-embedding-3-small",
    "litellm": "text-embedding-3-small",
    "local": "text-embedding-3-small",
    "gemini": "text-embedding-004",
    "ollama": "nomic-embed-text",
}


class EmbeddingProviderError(RuntimeError):
    pass


class NoEmbeddingProviderError(EmbeddingProviderError):
    pass


class EmbeddingsService:
    def __init__(self, db: Session) -> None:
        self.db = db

    def generate_embedding(self, user_id: str, text: str) -> list[float]:
        providers = self._list_embedding_providers(user_id)
        if not providers:
            raise NoEmbeddingProviderError(
                "No supported embedding provider is enabled. Add an OpenAI, Gemini, Ollama, OpenRouter, LiteLLM, or local OpenAI-compatible provider to enable retrieval."
            )
        return self._generate_embedding_with_providers(providers, text)

    def replace_chunks_for_source(
        self,
        *,
        user_id: str,
        source_type: str,
        source_id: str,
        asset_type: str,
        filename: str,
        chunks: list[tuple[str, dict]],
    ) -> list[AssetChunk]:
        (
            self.db.query(AssetChunk)
            .filter(
                AssetChunk.user_id == user_id,
                AssetChunk.source_type == source_type,
                AssetChunk.source_id == source_id,
            )
            .delete(synchronize_session=False)
        )

        providers = self._list_embedding_providers(user_id)
        if not providers:
            logger.warning("Skipping vector indexing for %s/%s because no embedding provider is configured.", source_type, source_id)
            self.db.flush()
            return []

        prepared_chunks: list[tuple[int, str, dict[str, Any], list[float]]] = []
        try:
            for chunk_index, (content, metadata) in enumerate(chunks):
                normalized = content.strip()
                if not normalized:
                    continue
                payload = dict(metadata or {})
                payload.setdefault("chunk_index", chunk_index)
                payload.setdefault("source_type", source_type)
                payload.setdefault("source_id", source_id)
                payload.setdefault("asset_type", asset_type)
                payload.setdefault("filename", filename)
                embedding = self._generate_embedding_with_providers(providers, normalized)
                prepared_chunks.append((chunk_index, normalized, payload, embedding))
        except EmbeddingProviderError as exc:
            logger.warning("Skipping vector indexing for %s/%s after embedding provider failure: %s", source_type, source_id, exc)
            self.db.flush()
            return []

        stored_chunks: list[AssetChunk] = []
        for chunk_index, normalized, payload, embedding in prepared_chunks:
            stored = AssetChunk(
                user_id=user_id,
                source_type=source_type,
                source_id=source_id,
                asset_type=asset_type,
                filename=filename,
                chunk_index=chunk_index,
                content=normalized,
                embedding=embedding,
                meta=payload,
            )
            self.db.add(stored)
            stored_chunks.append(stored)

        self.db.flush()
        return stored_chunks

    def delete_chunks_for_source(self, *, user_id: str, source_type: str, source_id: str) -> int:
        count = (
            self.db.query(AssetChunk)
            .filter(
                AssetChunk.user_id == user_id,
                AssetChunk.source_type == source_type,
                AssetChunk.source_id == source_id,
            )
            .delete(synchronize_session=False)
        )
        self.db.flush()
        return count

    def _list_embedding_providers(self, user_id: str) -> list[Provider]:
        workspace = WorkspaceService(self.db).get_or_create()
        query = self.db.query(Provider).filter(
            Provider.is_enabled.is_(True),
            Provider.provider_type.in_(SUPPORTED_EMBEDDING_PROVIDER_TYPES),
        )
        if workspace.authentication_enabled:
            query = query.filter(Provider.user_id == user_id)
        providers = list(query.order_by(Provider.created_at.asc()).all())
        providers.sort(key=lambda provider: EMBEDDING_PROVIDER_PRIORITY.get(provider.provider_type, 100))
        return providers

    def _generate_embedding_with_providers(self, providers: list[Provider], text: str) -> list[float]:
        errors: list[str] = []
        for provider in providers:
            try:
                vector = self._embed_with_provider(provider, text or " ")
                return self._coerce_embedding_dimension(vector)
            except EmbeddingProviderError as exc:
                errors.append(f"{provider.name}: {exc}")

        raise EmbeddingProviderError(errors[-1] if errors else "Embedding generation failed.")

    def _embed_with_provider(self, provider: Provider, text: str) -> list[float]:
        if provider.provider_type in OPENAI_COMPATIBLE_PROVIDER_TYPES:
            return self._embed_with_openai_compatible(provider, text)
        if provider.provider_type == "gemini":
            return self._embed_with_gemini(provider, text)
        if provider.provider_type == "ollama":
            return self._embed_with_ollama(provider, text)
        raise EmbeddingProviderError(f"{provider.provider_type} does not support embeddings.")

    def _embed_with_openai_compatible(self, provider: Provider, text: str) -> list[float]:
        encrypted_api_key = provider.encrypted_config.get("api_key")
        if provider.provider_type not in {"litellm", "local"} and not encrypted_api_key:
            raise EmbeddingProviderError(f"{provider.name} is missing an API key.")

        api_key = secret_manager.decrypt(encrypted_api_key) if encrypted_api_key else ""
        base_url = (provider.base_url or DEFAULT_BASE_URLS[provider.provider_type]).rstrip("/")
        payload = {
            "model": self._embedding_model_for_provider(provider),
            "input": text or " ",
        }
        headers = {"Content-Type": "application/json"}
        if api_key:
            headers["Authorization"] = f"Bearer {api_key}"

        try:
            with httpx.Client(timeout=60.0) as client:
                response = client.post(f"{base_url}/embeddings", headers=headers, json=payload)
                response.raise_for_status()
        except httpx.HTTPStatusError as exc:
            detail = exc.response.text or "Embedding request failed."
            raise EmbeddingProviderError(detail) from exc
        except httpx.HTTPError as exc:
            raise EmbeddingProviderError("Could not reach the embedding provider.") from exc

        data = response.json().get("data") or []
        if not data:
            raise EmbeddingProviderError("The embedding provider returned no vectors.")
        return self._parse_embedding_values(data[0].get("embedding"))

    def _embed_with_gemini(self, provider: Provider, text: str) -> list[float]:
        encrypted_api_key = provider.encrypted_config.get("api_key")
        if not encrypted_api_key:
            raise EmbeddingProviderError(f"{provider.name} is missing an API key.")

        api_key = secret_manager.decrypt(encrypted_api_key)
        base_url = (provider.base_url or DEFAULT_BASE_URLS["gemini"]).rstrip("/")
        model = self._embedding_model_for_provider(provider)
        model_path = model if model.startswith("models/") else f"models/{model}"

        try:
            with httpx.Client(timeout=60.0) as client:
                response = client.post(
                    f"{base_url}/{model_path}:embedContent",
                    params={"key": api_key},
                    json={"content": {"parts": [{"text": text or " "}]}},
                )
                response.raise_for_status()
        except httpx.HTTPStatusError as exc:
            detail = exc.response.text or "Embedding request failed."
            raise EmbeddingProviderError(detail) from exc
        except httpx.HTTPError as exc:
            raise EmbeddingProviderError("Could not reach the embedding provider.") from exc

        embedding = response.json().get("embedding") or {}
        return self._parse_embedding_values(embedding.get("values"))

    def _embed_with_ollama(self, provider: Provider, text: str) -> list[float]:
        base_url = (provider.base_url or DEFAULT_BASE_URLS["ollama"]).rstrip("/")

        try:
            with httpx.Client(timeout=60.0) as client:
                response = client.post(
                    f"{base_url}/api/embed",
                    json={"model": self._embedding_model_for_provider(provider), "input": text or " "},
                )
                response.raise_for_status()
        except httpx.HTTPStatusError as exc:
            detail = exc.response.text or "Embedding request failed."
            raise EmbeddingProviderError(detail) from exc
        except httpx.HTTPError as exc:
            raise EmbeddingProviderError("Could not reach the embedding provider.") from exc

        payload = response.json()
        if isinstance(payload.get("embedding"), list):
            return self._parse_embedding_values(payload["embedding"])
        embeddings = payload.get("embeddings") or []
        if embeddings:
            return self._parse_embedding_values(embeddings[0])
        raise EmbeddingProviderError("The embedding provider returned no vectors.")

    def _embedding_model_for_provider(self, provider: Provider) -> str:
        configured = provider.encrypted_config.get("embedding_model")
        if isinstance(configured, str) and configured.strip():
            return configured.strip()
        return DEFAULT_EMBEDDING_MODELS[provider.provider_type]

    def _parse_embedding_values(self, values: Any) -> list[float]:
        if not isinstance(values, list):
            raise EmbeddingProviderError("The embedding provider returned an invalid vector payload.")
        vector = [float(value) for value in values if isinstance(value, (int, float))]
        if not vector:
            raise EmbeddingProviderError("The embedding provider returned an empty vector.")
        return vector

    def _coerce_embedding_dimension(self, vector: list[float]) -> list[float]:
        target_dimension = settings.embedding_dimension
        if len(vector) == target_dimension:
            return vector
        if len(vector) > target_dimension:
            return vector[:target_dimension]
        # Keep pgvector writes stable while providers return different native dimensions.
        return [*vector, *([0.0] * (target_dimension - len(vector)))]

File: frontend/core/embeddings/embeddingsService.ts
import { and, eq, inArray } from "drizzle-orm";

import { WorkspaceService } from "../auth/workspaceService.js";
import { getConfig } from "../config/index.js";
import { getSecretManager } from "../crypto/index.js";
import type { ChikaimaDatabase } from "../db/client.js";
import { providers } from "../db/schema.js";
import { VectorStore, type StoredChunk } from "../rag/vectorStore.js";

const OPENAI_COMPATIBLE_PROVIDER_TYPES = ["openai", "openrouter", "litellm", "local"] as const;
const SUPPORTED_EMBEDDING_PROVIDER_TYPES = [...OPENAI_COMPATIBLE_PROVIDER_TYPES, "gemini", "ollama"] as const;
const EMBEDDING_PROVIDER_PRIORITY: Record<string, number> = { openai: 0, openrouter: 1, gemini: 2, ollama: 3, litellm: 4 };

const DEFAULT_BASE_URLS: Record<string, string> = {
  openai: "https://api.openai.com/v1",
  openrouter: "https://openrouter.ai/api/v1",
  litellm: "http://localhost:4000/v1",
  local: "http://localhost:4000/v1",
  gemini: "https://generativelanguage.googleapis.com/v1beta",
  ollama: "http://localhost:11434",
};

const DEFAULT_EMBEDDING_MODELS: Record<string, string> = {
  openai: "text-embedding-3-small",
  openrouter: "openai/text-embedding-3-small",
  litellm: "text-embedding-3-small",
  local: "text-embedding-3-small",
  gemini: "text-embedding-004",
  ollama: "nomic-embed-text",
};

type ProviderRow = typeof providers.$inferSelect;

export class EmbeddingProviderError extends Error {}
export class NoEmbeddingProviderError extends EmbeddingProviderError {}

export class EmbeddingsService {
  private readonly vectorStore: VectorStore;

  constructor(private readonly db: ChikaimaDatabase) {
    this.vectorStore = new VectorStore(db);
  }

  async generateEmbedding(userId: string, text: string): Promise<number[]> {
    const candidateProviders = this.listEmbeddingProviders(userId);
    if (candidateProviders.length === 0) {
      throw new NoEmbeddingProviderError(
        "No supported embedding provider is enabled. Add an OpenAI, Gemini, Ollama, OpenRouter, LiteLLM, or local OpenAI-compatible provider to enable retrieval.",
      );
    }
    return this.generateEmbeddingWithProviders(candidateProviders, text);
  }

  async replaceChunksForSource(params: {
    userId: string;
    sourceType: string;
    sourceId: string;
    assetType: string;
    filename: string;
    chunks: Array<{ content: string; metadata: Record<string, unknown> }>;
  }): Promise<StoredChunk[]> {
    this.vectorStore.deleteBySource(params.userId, params.sourceType, params.sourceId);

    const candidateProviders = this.listEmbeddingProviders(params.userId);
    if (candidateProviders.length === 0) {
      return [];
    }

    const prepared: Array<{ content: string; metadata: Record<string, unknown>; embedding: number[] }> = [];
    try {
      for (let index = 0; index < params.chunks.length; index += 1) {
        const normalized = params.chunks[index]!.content.trim();
        if (!normalized) continue;

        const payload: Record<string, unknown> = {
          chunk_index: index,
          source_type: params.sourceType,
          source_id: params.sourceId,
          asset_type: params.assetType,
          filename: params.filename,
          ...params.chunks[index]!.metadata,
        };
        const embedding = await this.generateEmbeddingWithProviders(candidateProviders, normalized);
        prepared.push({ content: normalized, metadata: payload, embedding });
      }
    } catch (error) {
      if (error instanceof EmbeddingProviderError) {
        return [];
      }
      throw error;
    }

    if (prepared.length === 0) return [];

    return this.vectorStore.insertChunks({
      userId: params.userId,
      sourceType: params.sourceType,
      sourceId: params.sourceId,
      assetType: params.assetType,
      filename: params.filename,
      chunks: prepared,
    });
  }

  deleteChunksForSource(userId: string, sourceType: string, sourceId: string): number {
    return this.vectorStore.deleteBySource(userId, sourceType, sourceId);
  }

  private listEmbeddingProviders(userId: string): ProviderRow[] {
    const workspace = new WorkspaceService(this.db).getOrCreate();
    let rows = this.db
      .select()
      .from(providers)
      .where(and(eq(providers.isEnabled, true), inArray(providers.providerType, [...SUPPORTED_EMBEDDING_PROVIDER_TYPES])))
      .all();

    if (workspace.authenticationEnabled) {
      rows = rows.filter((provider) => provider.userId === userId);
    }
    rows.sort((a, b) => a.createdAt.localeCompare(b.createdAt));
    rows.sort((a, b) => (EMBEDDING_PROVIDER_PRIORITY[a.providerType] ?? 100) - (EMBEDDING_PROVIDER_PRIORITY[b.providerType] ?? 100));
    return rows;
  }

  private async generateEmbeddingWithProviders(candidateProviders: ProviderRow[], text: string): Promise<number[]> {
    const errors: string[] = [];
    for (const provider of candidateProviders) {
      try {
        const vector = await this.embedWithProvider(provider, text || " ");
        return this.coerceEmbeddingDimension(vector);
      } catch (error) {
        if (error instanceof EmbeddingProviderError) {
          errors.push(`${provider.name}: ${error.message}`);
        } else {
          throw error;
        }
      }
    }
    throw new EmbeddingProviderError(errors[errors.length - 1] ?? "Embedding generation failed.");
  }

  private async embedWithProvider(provider: ProviderRow, text: string): Promise<number[]> {
    if ((OPENAI_COMPATIBLE_PROVIDER_TYPES as readonly string[]).includes(provider.providerType)) {
      return this.embedWithOpenaiCompatible(provider, text);
    }
    if (provider.providerType === "gemini") {
      return this.embedWithGemini(provider, text);
    }
    if (provider.providerType === "ollama") {
      return this.embedWithOllama(provider, text);
    }
    throw new EmbeddingProviderError(`${provider.providerType} does not support embeddings.`);
  }

  private async embedWithOpenaiCompatible(provider: ProviderRow, text: string): Promise<number[]> {
    const config = provider.encryptedConfig as Record<string, unknown>;
    const encryptedApiKey = config.api_key;
    if (!["litellm", "local"].includes(provider.providerType) && typeof encryptedApiKey !== "string") {
      throw new EmbeddingProviderError(`${provider.name} is missing an API key.`);
    }
    const apiKey = typeof encryptedApiKey === "string" ? getSecretManager().decrypt(encryptedApiKey) : "";
    const baseUrl = (provider.baseUrl || DEFAULT_BASE_URLS[provider.providerType]!).replace(/\/+$/, "");

    const headers: Record<string, string> = { "Content-Type": "application/json" };
    if (apiKey) headers.Authorization = `Bearer ${apiKey}`;

    const response = await this.postJson(`${baseUrl}/embeddings`, headers, {
      model: this.embeddingModelForProvider(provider),
      input: text || " ",
    });
    const data = (response.data as Array<{ embedding?: unknown }> | undefined) ?? [];
    if (data.length === 0) {
      throw new EmbeddingProviderError("The embedding provider returned no vectors.");
    }
    return this.parseEmbeddingValues(data[0]!.embedding);
  }

  private async embedWithGemini(provider: ProviderRow, text: string): Promise<number[]> {
    const config = provider.encryptedConfig as Record<string, unknown>;
    const encryptedApiKey = config.api_key;
    if (typeof encryptedApiKey !== "string") {
      throw new EmbeddingProviderError(`${provider.name} is missing an API key.`);
    }
    const apiKey = getSecretManager().decrypt(encryptedApiKey);
    const baseUrl = (provider.baseUrl || DEFAULT_BASE_URLS.gemini!).replace(/\/+$/, "");
    const model = this.embeddingModelForProvider(provider);
    const modelPath = model.startsWith("models/") ? model : `models/${model}`;

    const response = await this.postJson(`${baseUrl}/${modelPath}:embedContent?key=${encodeURIComponent(apiKey)}`, {}, { content: { parts: [{ text: text || " " }] } });
    const embedding = (response.embedding as { values?: unknown } | undefined) ?? {};
    return this.parseEmbeddingValues(embedding.values);
  }

  private async embedWithOllama(provider: ProviderRow, text: string): Promise<number[]> {
    const baseUrl = (provider.baseUrl || DEFAULT_BASE_URLS.ollama!).replace(/\/+$/, "");
    const response = await this.postJson(`${baseUrl}/api/embed`, {}, { model: this.embeddingModelForProvider(provider), input: text || " " });

    if (Array.isArray(response.embedding)) {
      return this.parseEmbeddingValues(response.embedding);
    }
    const embeddings = (response.embeddings as unknown[] | undefined) ?? [];
    if (embeddings.length > 0) {
      return this.parseEmbeddingValues(embeddings[0]);
    }
    throw new EmbeddingProviderError("The embedding provider returned no vectors.");
  }

  private async postJson(url: string, headers: Record<string, string>, body: unknown): Promise<Record<string, unknown>> {
    let response: Response;
    try {
      response = await fetch(url, { method: "POST", headers: { "Content-Type": "application/json", ...headers }, body: JSON.stringify(body) });
    } catch {
      throw new EmbeddingProviderError("Could not reach the embedding provider.");
    }
    if (!response.ok) {
      const detail = (await response.text().catch(() => "")).trim();
      throw new EmbeddingProviderError(detail || "Embedding request failed.");
    }
    return (await response.json()) as Record<string, unknown>;
  }

  private embeddingModelForProvider(provider: ProviderRow): string {
    const config = provider.encryptedConfig as Record<string, unknown>;
    const configured = config.embedding_model;
    if (typeof configured === "string" && configured.trim()) {
      return configured.trim();
    }
    return DEFAULT_EMBEDDING_MODELS[provider.providerType]!;
  }

  private parseEmbeddingValues(values: unknown): number[] {
    if (!Array.isArray(values)) {
      throw new EmbeddingProviderError("The embedding provider returned an invalid vector payload.");
    }
    const vector = values.filter((value): value is number => typeof value === "number").map((value) => Number(value));
    if (vector.length === 0) {
      throw new EmbeddingProviderError("The embedding provider returned an empty vector.");
    }
    return vector;
  }

  private coerceEmbeddingDimension(vector: number[]): number[] {
    const target = getConfig().embeddingDimension;
    if (vector.length === target) return vector;
    if (vector.length > target) return vector.slice(0, target);
    return [...vector, ...new Array(target - vector.length).fill(0)];
  }
}
